document/FileFormat.mdk

```diff
@@ -155,14 +155,41 @@ property, for example, "this is a source file".
 5. Brevity: The URI base id property might be shorter than the absolute path it represents.
 ~
 
+## String properties { #string-properties }
+
+Unless otherwise specified in the description of a specific property,
+all properties whose values are of type `"string"` must have a non-empty value.
+
+## Object properties { #object-properties }
+
+Certain properties in this specification are defined to be JSON objects whose property names
+satisfy certain conditions.
+Examples are the `run.files` property (&sect;[#run-files]) and the
+`rule.messageFormats` property (&sect;[#rule-messageFormats]).
+Unless otherwise specified in the description of a specific property,
+if any such object is empty,
+then the either property may be represented as an empty object `{}`,
+or it may be absent.
+
+## Array properties { #array-properties }
+
+Certain properties in this specification are defined to be JSON arrays.
+Examples are the `run.toolNotifications` property (&sect;[#run-toolNotifications]) and the
+`file.hash` property (&sect;[#file-hashes]).
```

Comment:
fixed.

---
In reply to: [75180701](https://github.com/sarif-standard/sarif-spec/pull/256#discussion_r75180701) [](ancestors = 75180701)